File: backend/app/schemas/crop.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel


class CropIn(BaseModel):
    plant_id: uuid.UUID | None = None
    name: str = Field(min_length=1, max_length=100)
    variety: str | None = None
    planting_date: date | None = None
    area: Decimal | None = Field(default=None, ge=0)
    location: str | None = None
    notes: str | None = None


class CropUpdate(BaseModel):
    plant_id: uuid.UUID | None = None
    name: str | None = Field(default=None, min_length=1, max_length=100)
    variety: str | None = None
    planting_date: date | None = None
    area: Decimal | None = Field(default=None, ge=0)
    location: str | None = None
    status: str | None = Field(default=None, pattern=r"^(active|archived)$")
    health_score: int | None = Field(default=None, ge=0, le=100)
    notes: str | None = None


class CropOut(ORMModel):
    id: uuid.UUID
    plant_id: uuid.UUID | None = None
    plant_name: str | None = None
    name: str
    variety: str | None = None
    planting_date: date | None = None
    area: Decimal | None = None
    location: str | None = None
    status: str
    health_score: int
    notes: str | None = None
    created_at: datetime | None = None
    last_disease: str | None = None


class CropLogIn(BaseModel):
    content: str = Field(min_length=1)
    image_url: str | None = None


class CropLogOut(ORMModel):
    id: uuid.UUID
    crop_id: uuid.UUID
    content: str
    image_url: str | None = None
    created_at: datetime | None = None


class HealthPoint(BaseModel):
    date: datetime
    health_score: int
    disease: str | None = None
    diagnosis_id: uuid.UUID | None = None

File: mobile/lib/domain/entities/crop.dart
import '../../core/utils/format.dart';

class Crop {
  Crop({
    required this.id,
    required this.name,
    this.plantId,
    this.plantName,
    this.variety,
    this.plantingDate,
    this.area,
    this.location,
    this.status = 'active',
    this.healthScore = 100,
    this.notes,
    this.lastDisease,
    this.createdAt,
  });
  final String id, name, status;
  final String? plantId, plantName, variety, location, notes, lastDisease;
  final DateTime? plantingDate, createdAt;
  final double? area;
  final int healthScore;

  factory Crop.fromJson(Map<String, dynamic> j) => Crop(
        id: j['id'],
        name: j['name'],
        plantId: j['plant_id'],
        plantName: j['plant_name'],
        variety: j['variety'],
        plantingDate: parseDate(j['planting_date']),
        area: j['area'] == null ? null : double.tryParse(j['area'].toString()),
        location: j['location'],
        status: j['status'] ?? 'active',
        healthScore: j['health_score'] ?? 100,
        notes: j['notes'],
        lastDisease: j['last_disease'],
        createdAt: parseDate(j['created_at']),
      );
}

class HealthPoint {
  HealthPoint(this.date, this.score, this.disease);
  final DateTime date;
  final int score;
  final String? disease;

  factory HealthPoint.fromJson(Map<String, dynamic> j) =>
      HealthPoint(parseDate(j['date']) ?? DateTime.now(), j['health_score'] ?? 0, j['disease']);
}

class CropLog {
  CropLog({required this.id, required this.content, this.imageUrl, this.createdAt});
  final String id, content;
  final String? imageUrl;
  final DateTime? createdAt;

  factory CropLog.fromJson(Map<String, dynamic> j) =>
      CropLog(id: j['id'], content: j['content'], imageUrl: j['image_url'], createdAt: parseDate(j['created_at']));
}
